### ISSUE TITLE
TypeScript non-null assertion operator not stripped off with deep optional chaining

### ISSUE THREAD

**Issue by maxdavidson** [babel/babel#12386]
## Bug Report

- [ ] I would like to work on a fix!

I noticed a really strange regression in in @babel/core 7.12.7. When used with optional chaining, the TypeScript postfix "!" operator (non-null assertion) sometimes won't get stripped off, resulting in a syntax error.

Input:
```js
const a = 1;
const b = () => a?.b?.c!.d
```

Output with [@babel/core@7.12.3](https://babel.dev/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBAhjAvDAjAbgFCkrARkmACgEokA-eAfgDpcbgBCagEwyA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=typescript&prettier=false&targets=&version=7.12.3&externalPlugins=%40babel%2Fplugin-proposal-optional-chaining%407.12.7
):
```js
const a = 1;

const b = () => {
  var _a$b;

  return a === null || a === void 0 ? void 0 : (_a$b = a.b) === null || _a$b === void 0 ? void 0 : _a$b.c.d;
};
```

Output with [@babel/core@7.12.7](https://babel.dev/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBAhjAvDAjAbgFCkrARkmACgEokA-eAfgDpcbgBCagEwyA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=typescript&prettier=false&targets=&version=7.12.7&externalPlugins=%40babel%2Fplugin-proposal-optional-chaining%407.12.7
):
```js
const a = 1;

const b = () => {
  var _a$b;

  return a === null || a === void 0 ? void 0 : (_a$b = a.b) === null || _a$b === void 0 ? void 0 : _a$b.c!.d;
};
```

Notice the `_a$b.c!.d` at the end

Plugins used:
- @babel/preset-typescript@7.12.7
- @babel/plugin-proposal-optional-chaining@7.12.7 (via preset-env)

**Comment by JLHwung** [comment]
Another minimal test [case](https://babeljs.io/repl/build/31959/#?browsers=safari%2012&build=&builtIns=false&spec=false&loose=false&code_lz=IYAgvAfCwPwHQCMCE8DGAoIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=true&sourceType=script&lineWrap=false&presets=env%2Ctypescript&prettier=false&targets=&version=7.12.3%2Bpr.12331&externalPlugins=):
```js
a => a?.b!?.c
```
This is likely a regression introduced at #12302.

**Comment by nicolo-ribaudo** [comment]
Fixed in `7.12.9`
